Nested Named ValueTuples throw exception

Similar to #241, but for nested generic definitions

When defining a shape like this:

```csharp
[GenerateShape]
public sealed partial record Message
{
    public required ImmutableDictionary<string, (ImmutableArray<int> A, ImmutableArray<uint> B)> C { get; init; }
}
```

The following error is produced on `PolyType` version `0.56.1`:

```text
warning CS8785: Generator 'PolyTypeGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'KeyNotFoundException' with message 'The given key 'global::System.Collections.Immutable.ImmutableDictionary<string, (global::System.Collections.Immutable.ImmutableArray<int> A, global::System.Collections.Immutable.ImmutableArray<uint> B)>' was not present in the dictionary.'.
```

It appears it has something to do with the fact that the tuple is named. If I remove the names, it works fine:
```csharp
[GenerateShape]
public sealed partial record Message
{
    public required ImmutableDictionary<string, (ImmutableArray<int>, ImmutableArray<uint>)> C { get; init; }
}
```
or:
```csharp
[GenerateShape]
public sealed partial record Message
{
    public required ImmutableDictionary<string, ValueTuple<ImmutableArray<int>, ImmutableArray<uint>>> C { get; init; }
}
```

--- a/src/PolyType.SourceGenerator/Helpers/RoslynHelpers.cs
+++ b/src/PolyType.SourceGenerator/Helpers/RoslynHelpers.cs
@@ -29,77 +29,93 @@ internal static partial class RoslynHelpers
     /// <summary>
     /// Removes erased compiler metadata such as tuple names and nullable annotations.
     /// </summary>
-    public static ITypeSymbol EraseCompilerMetadata(this Compilation compilation, ITypeSymbol type)
+    public static ITypeSymbol EraseCompilerMetadata(this Compilation compilation, ITypeSymbol type, bool useForSymbolDisplayOnly = false)
     {
-        if (!RequiresErasure(type))
+        if (useForSymbolDisplayOnly && !SymbolDisplayRequiresErasure(type))
         {
             return type;
         }
 
-        if (type.NullableAnnotation != NullableAnnotation.None)
-        {
-            type = type.WithNullableAnnotation(NullableAnnotation.None);
-        }
+        return EraseCore(type);
 
-        if (type is IArrayTypeSymbol arrayType)
+        ITypeSymbol EraseCore(ITypeSymbol type)
         {
-            ITypeSymbol elementType = compilation.EraseCompilerMetadata(arrayType.ElementType);
-            return compilation.CreateArrayTypeSymbol(elementType, arrayType.Rank);
-        }
-
-        if (type is INamedTypeSymbol namedType)
-        {
-            if (namedType is { IsTupleType: true, TupleElements.Length: >= 2 })
+            if (type.NullableAnnotation is not NullableAnnotation.None)
             {
-                ImmutableArray<ITypeSymbol> erasedElements = namedType.TupleElements
-                    .Select(e => compilation.EraseCompilerMetadata(e.Type))
-                    .ToImmutableArray();
-
-                type = compilation.CreateTupleTypeSymbol(erasedElements);
+                type = type.WithNullableAnnotation(NullableAnnotation.None);
             }
-            else if (namedType.IsGenericType)
-            {
-                ImmutableArray<ITypeSymbol> typeArguments = namedType.TypeArguments;
-                INamedTypeSymbol? containingType = namedType.ContainingType;
 
-                if (containingType?.IsGenericType == true)
-                {
-                    containingType = (INamedTypeSymbol)compilation.EraseCompilerMetadata(containingType);
-                    type = namedType = containingType.GetTypeMembers().First(t => t.Name == namedType.Name && t.Arity == namedType.Arity);
-                }
-
-                if (typeArguments.Length > 0)
-                {
-                    ITypeSymbol[] erasedTypeArgs = typeArguments
-                        .Select(compilation.EraseCompilerMetadata)
-                        .ToArray();
-
-                    type = namedType.ConstructedFrom.Construct(erasedTypeArgs);
-                }
-            }
-        }
-
-        return type;
-
-        static bool RequiresErasure(ITypeSymbol type)
-        {
-            if (type.NullableAnnotation != NullableAnnotation.None)
-            {
-                return true;
-            }
             if (type is IArrayTypeSymbol arrayType)
             {
-                return RequiresErasure(arrayType.ElementType);
+                ITypeSymbol elementType = compilation.EraseCompilerMetadata(arrayType.ElementType);
+                return compilation.CreateArrayTypeSymbol(elementType, arrayType.Rank);
+            }
+
+            if (type is INamedTypeSymbol namedType)
+            {
+                if (namedType is { IsTupleType: true, TupleElements.Length: >= 2 })
+                {
+                    ImmutableArray<ITypeSymbol> erasedElements = namedType.TupleElements
+                        .Select(e => compilation.EraseCompilerMetadata(e.Type))
+                        .ToImmutableArray();
+
+                    type = compilation.CreateTupleTypeSymbol(erasedElements);
+                }
+                else if (namedType.IsGenericType)
+                {
+                    ImmutableArray<ITypeSymbol> typeArguments = namedType.TypeArguments;
+                    INamedTypeSymbol? containingType = namedType.ContainingType;
+
+                    if (containingType?.IsGenericType is true)
+                    {
+                        containingType = (INamedTypeSymbol)EraseCore(containingType);
+                        type = namedType = containingType.GetTypeMembers().First(t => t.Name == namedType.Name && t.Arity == namedType.Arity);
+                    }
+
+                    if (typeArguments.Length > 0)
+                    {
+                        ITypeSymbol[] erasedTypeArgs = typeArguments
+                            .Select(EraseCore)
+                            .ToArray();
+
+                        type = namedType.ConstructedFrom.Construct(erasedTypeArgs);
+                    }
+                }
+            }
+
+            return type;
+        }
+
+        static bool SymbolDisplayRequiresErasure(ITypeSymbol type)
+        {
+            // No need to check for type.NullableAnnotation since SymbolDisplayFormat.FullyQualifiedFormat
+            // does not enable IncludeNullableReferenceTypeModifier.
+
+            if (type is IArrayTypeSymbol arrayType)
+            {
+                return SymbolDisplayRequiresErasure(arrayType.ElementType);
             }
             if (type is INamedTypeSymbol namedType)
             {
                 if (namedType.IsTupleType)
                 {
-                    return namedType.TupleElements.Any(e => RequiresErasure(e.Type));
+                    foreach (IFieldSymbol e in namedType.TupleElements)
+                    {
+                        if (!string.IsNullOrEmpty(e.Name) || SymbolDisplayRequiresErasure(e.Type))
+                        {
+                            return true;
+                        }
+                    }
                 }
-                if (namedType.TypeArguments.Length > 0)
+                else if (namedType.TypeArguments.Length > 0)
                 {
-                    return namedType.TypeArguments.Any(RequiresErasure);
+                    foreach (ITypeSymbol t in namedType.TypeArguments)
+                    {
+                        if (SymbolDisplayRequiresErasure(t))
+                        {
+                            return true;
+                        }
+                    }
                 }
             }
 

--- a/src/PolyType.SourceGenerator/Helpers/RoslynHelpers.cs
+++ b/src/PolyType.SourceGenerator/Helpers/RoslynHelpers.cs
@@ -31,6 +31,11 @@ internal static partial class RoslynHelpers
     /// </summary>
     public static ITypeSymbol EraseCompilerMetadata(this Compilation compilation, ITypeSymbol type)
     {
+        if (!RequiresErasure(type))
+        {
+            return type;
+        }
+
         if (type.NullableAnnotation != NullableAnnotation.None)
         {
             type = type.WithNullableAnnotation(NullableAnnotation.None);
@@ -44,13 +49,8 @@ internal static partial class RoslynHelpers
 
         if (type is INamedTypeSymbol namedType)
         {
-            if (namedType.IsTupleType)
+            if (namedType is { IsTupleType: true, TupleElements.Length: >= 2 })
             {
-                if (namedType.TupleElements.Length < 2)
-                {
-                    return type;
-                }
-
                 ImmutableArray<ITypeSymbol> erasedElements = namedType.TupleElements
                     .Select(e => compilation.EraseCompilerMetadata(e.Type))
                     .ToImmutableArray();
@@ -80,6 +80,31 @@ internal static partial class RoslynHelpers
         }
 
         return type;
+
+        static bool RequiresErasure(ITypeSymbol type)
+        {
+            if (type.NullableAnnotation != NullableAnnotation.None)
+            {
+                return true;
+            }
+            if (type is IArrayTypeSymbol arrayType)
+            {
+                return RequiresErasure(arrayType.ElementType);
+            }
+            if (type is INamedTypeSymbol namedType)
+            {
+                if (namedType.IsTupleType)
+                {
+                    return namedType.TupleElements.Any(e => RequiresErasure(e.Type));
+                }
+                if (namedType.TypeArguments.Length > 0)
+                {
+                    return namedType.TypeArguments.Any(RequiresErasure);
+                }
+            }
+
+            return false;
+        }
     }
 
     public static bool ContainsNullabilityAnnotations(this ITypeSymbol type)

--- a/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
@@ -338,7 +338,7 @@ public sealed partial class Parser
                 _ => false
             },
 
-            PropertyType = CreateTypeId(NormalizeType(property.PropertyType)),
+            PropertyType = CreateTypeId(property.PropertyType),
             IsGetterNonNullable = emitGetter && property.IsGetterNonNullable,
             IsSetterNonNullable = emitSetter && property.IsSetterNonNullable,
             PropertyTypeContainsNullabilityAnnotations = property.PropertyType.ContainsNullabilityAnnotations(),
@@ -571,7 +571,7 @@ public sealed partial class Parser
             _ => underlyingValue.ToString(),
         };
 
-    private static ConstructorShapeModel MapTupleConstructor(TypeId typeId, TupleDataModel tupleModel)
+    private ConstructorShapeModel MapTupleConstructor(TypeId typeId, TupleDataModel tupleModel)
     {
         return new ConstructorShapeModel
         {
@@ -593,7 +593,7 @@ public sealed partial class Parser
             IsPublic = true,
         };
 
-        static ParameterShapeModel MapTupleConstructorParameter(TypeId typeId, PropertyDataModel tupleElement, int position)
+        ParameterShapeModel MapTupleConstructorParameter(TypeId typeId, PropertyDataModel tupleElement, int position)
         {
             string name = $"Item{position + 1}";
             return new ParameterShapeModel

--- a/src/PolyType.SourceGenerator/Parser/Parser.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.cs
@@ -1051,7 +1051,7 @@ public sealed partial class Parser : TypeDataModelGenerator
 
     private TypeId CreateTypeId(ITypeSymbol type)
     {
-        type = NormalizeType(type);
+        type = KnownSymbols.Compilation.EraseCompilerMetadata(type, useForSymbolDisplayOnly: true);
         return new TypeId
         {
             FullyQualifiedName = type.GetFullyQualifiedName(),

--- a/src/PolyType.SourceGenerator/Parser/Parser.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.cs
@@ -973,7 +973,7 @@ public sealed partial class Parser : TypeDataModelGenerator
         }
     }
 
-    private static AssociatedTypeId CreateAssociatedTypeId(INamedTypeSymbol open, INamedTypeSymbol closed)
+    private AssociatedTypeId CreateAssociatedTypeId(INamedTypeSymbol open, INamedTypeSymbol closed)
     {
         StringBuilder builder = new();
 
@@ -1049,8 +1049,9 @@ public sealed partial class Parser : TypeDataModelGenerator
         }
     }
 
-    private static TypeId CreateTypeId(ITypeSymbol type)
+    private TypeId CreateTypeId(ITypeSymbol type)
     {
+        type = NormalizeType(type);
         return new TypeId
         {
             FullyQualifiedName = type.GetFullyQualifiedName(),

--- a/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.cs
+++ b/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.cs
@@ -1429,8 +1429,33 @@ public static partial class CompilationTests
                 public (string A, string B) Foo { get; set; }
                 public List<(string X, int Y)>? Bar { get; set; }
                 public Dictionary<(int I, string S), (string U, string V)>? Baz { get; set; }
+                public (string A, (string X, int Y)) Qux { get; set; }
             }
             """);
+
+        PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
+        Assert.Empty(result.Diagnostics);
+    }
+
+    [Fact]
+    public static void NamedTupleParameters()
+    {
+        // Regression test for https://github.com/eiriktsarpalis/PolyType/issues/244
+        Compilation compilation = CompilationHelpers.CreateCompilation("""
+            using PolyType;
+            using System.Collections.Generic;
+
+            [GenerateShape]
+            public partial record Message(
+                (string A, string B) foo,
+                List<(string X, int Y)> bar,
+                Dictionary<(int I, string S), (string U, string V)> baz,
+                (string A, (string X, int Y)) qux)
+            {
+                public required (string A, string B) RequiredProp { get; set; }
+            }
+            """,
+            parseOptions: new(LanguageVersion.CSharp12));
 
         PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
         Assert.Empty(result.Diagnostics);

